ecs-agent/acs/session/payload_responder.go

```diff
@@ -0,0 +1,162 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package session
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+)
+
+const (
+	PayloadMessageName = "PayloadMessage"
+)
+
+type TaskMessageHandler interface {
+	ProcessTaskMessage(message *ecsacs.PayloadMessage,
+		ackFunc func(*ecsacs.AckRequest, []*ecsacs.IAMRoleCredentialsAckRequest)) error
+}
+
+// payloadResponder implements the wsclient.RequestResponder interface for responding
+// to ecsacs.PayloadMessage messages sent by ACS.
+type payloadResponder struct {
+	taskMessageHandler TaskMessageHandler
+	respond            wsclient.RespondFunc
+}
+
+// NewPayloadResponder returns an instance of the payloadResponder struct.
+func NewPayloadResponder(taskMessageHandler TaskMessageHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &payloadResponder{
+		taskMessageHandler: taskMessageHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*payloadResponder) Name() string { return "payload responder" }
+
+func (r *payloadResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handlePayloadMessage
+}
+
+func (r *payloadResponder) handlePayloadMessage(message *ecsacs.PayloadMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", PayloadMessageName))
+
+	// Validate fields in the message.
+	if err := validatePayloadMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", PayloadMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle payload message.
+	err := r.taskMessageHandler.ProcessTaskMessage(message, r.ackFunc)
+	if err != nil {
+		logger.Critical(fmt.Sprintf("Unable to handle %s", PayloadMessageName), logger.Fields{
+			field.MessageID: aws.StringValue(message.MessageId),
+			field.Error:     err,
+		})
+	}
+}
+
+// ackFunc sends ACKs of the payload message and of the credentials associated with the tasks contained in the payload
+// message.
+func (r *payloadResponder) ackFunc(payloadAck *ecsacs.AckRequest, credsAcks []*ecsacs.IAMRoleCredentialsAckRequest) {
+	go r.sendAck(payloadAck)
```

Comment:
Since, we are calling these asynchronously, should we not call them with cancellable context?

Reply:
I don't think calling with context is necessary here, we don't require a context when sending simple ACKs to ACS right now and I don't see any common reasons for needing to call with context applying here. Is there anything specific to you that seems like calling with a context would be required here?

Reply:
I suppose it is related to retry mechanism. When acs has connection issue would it be possible to stay forever to retry causing resource leak? 

Also having it will help with agent exit case (to cancel all pending calls). This is a suggestion from Anirudh during TACS design review
> Can we add it to our developer guidelines doc (if it's not there already) that we should never create go routines without a cancelable context and that folks should only create a background context in the main method?

Reply:
I think if the call itself is short-living (and/or no retries are needed and implemented) it is not a big deal but I think we would still want it if possible.

Reply:
> Also having it will help with agent exit case (to cancel all pending calls). This is a suggestion from Anirudh during TACS design review

> I think if the call itself is short-living (and/or no retries are needed and implemented) it is not a big deal but I think we would still want it if possible.

Thanks for sharing this guidance, I was not aware of this. I think there may be a larger issue as there are many other places in ecs-agent module right now where we create goroutines without a cancelable context, some of which are NOT short-lived and/or implement retries.

We currently do not pass in the main method's context to ACS message responders so it would be difficult/require a large refactor to achieve this.

> When acs has connection issue would it be possible to stay forever to retry causing resource leak

Going through the code trace, I don't think it's possible for such calls to stay forever. They are short-lived and no retries are implemented.

Given the above reasons, I think we can call this out of scope of this PR - but I will take an action item to follow up with Anirudh and/or the team to evaluate where we currently create goroutines without a cancelable context, especially goroutines that are NOT short-lived and/or implement retries (unlike in this case) and discuss what actions (if any) should be done about them. Does this sound good to you?

Reply:
Sounds fine to me to call it out of scope